src/axom/core/ArrayBase.hpp

```diff
@@ -10,14 +10,38 @@
 #include "axom/core/Macros.hpp"               // for axom macros
 #include "axom/core/utilities/Utilities.hpp"  // for processAbort()
 #include "axom/core/Types.hpp"                // for IndexType definition
+#include "axom/core/StackArray.hpp"
+#include "axom/core/numerics/matvecops.hpp"  // for dot_product
 
 // C/C++ includes
-#include <array>     // for std::array
 #include <iostream>  // for std::cerr and std::ostream
-#include <numeric>   // for std::inner_product
+#include <numeric>   // for std::accumulate
+
+#ifdef AXOM_USE_UMPIRE
+  #include "umpire/resource/MemoryResourceTypes.hpp"
+#endif
 
 namespace axom
 {
+/*! 
+ * \brief Memory spaces supported by Array-like types
+ *
+ * This abstraction is not implemented using Umpire's MemoryResourceType enum
+ * in order to also include a "Dynamic" option as a default template parameter
+ * for Array-like types
+ */
+enum class MemorySpace
+{
+  Dynamic,
+#ifdef AXOM_USE_UMPIRE
+  Host,
+  Device,
+  Unified,
+  Pinned,
+  Constant
```

Comment:
Out of curiosity - should we enforce a compile-time constraint that `Array`s or `ArrayView`s are read-only on the GPU for `MemorySpace::Constant`?

Reply:
This should be possible via a `static_assert` in the non-const `data()` accessor in `ArrayView` - I'll go ahead and add that in

Reply:
Does having something like:
```c++
#ifdef AXOM_DEVICE_CODE
  static_assert(SPACE == MemorySpace::Constant);
#endif
```
in the non-const overloads of `Array::data()` and `Array::operator[]` not work? I think that should allow you to write to the array from the host.

Reply:
I think that would still restrict host usage of that function if the corresponding host code was compiled with NVCC - that is, I think that macro is defined for both the host and device passes.

Reply:
IIRC, I think `AXOM_DEVICE_CODE` is equivalent to `__CUDA_ARCH__`, which should only be defined on the device code pass.